src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/join/JoinRewriteRule.java

```diff
@@ -0,0 +1,191 @@
+        return query.getFrom() instanceof SQLJoinTableSource
+            && ((SQLJoinTableSource) query.getFrom()).getJoinType() != SQLJoinTableSource.JoinType.COMMA;
+    }
+
+    @Override
+    public void rewrite(SQLQueryExpr root) {
+
+        final Multimap<String, Table> tableByFieldName = ArrayListMultimap.create();
+        final Set<String> tableNameAndAlias = new HashSet<>();
+        final Map<String, String> tableNameToAlias = new HashMap<>();
+
+        visitTable(root, tableExpr -> {
+            // Copied from SubqueryAliasRewriter ; Removes index type name if any
+            String tableName = tableExpr.getExpr().toString().replaceAll(" ", "").split("/")[0];
+
+            if (tableExpr.getAlias() == null) {
+                // Could we not directly use table name as alias?
+                tableExpr.setAlias(createAlias(tableName));
+            }
+
+            Table table = new Table(tableName, tableExpr.getAlias());
+
+            tableNameAndAlias.add(table.getName());
+            tableNameAndAlias.add(table.getAlias());
+            tableNameToAlias.put(table.getName(), table.getAlias());
+
+            FieldMappings fieldMappings = clusterState.getFieldMappings(
+                new String[]{tableName}).firstMapping().firstMapping();
+            fieldMappings.flat((fieldName, type) -> tableByFieldName.put(fieldName, table));
+        });
+
+        visitColumnName(root, idExpr -> {
+            String columnName = idExpr.getName();
+            Collection<Table> tables = tableByFieldName.get(columnName);
+            if (tables.size() > 1) {
+                // columnName without alias present in both tables
+                throw new VerificationException(StringUtils.format("Field name [%s] is ambiguous", columnName));
+            } else if (tables.isEmpty()) {
+                // size() == 0?
+                // 1. Either the columnName does not exist (handled by SemanticAnalyzer [SemanticAnalysisException])
+                // 2. Or column starts with tableName as alias or explicit alias
+                //    If starts with tableName as alias change to explicit alias
+                tableNameToAlias.keySet().stream().forEach(tableName -> {
+                    if (columnName.startsWith(tableName + DOT)) {
+                        idExpr.setName(columnName.replace(tableName + DOT, tableNameToAlias.get(tableName) + DOT));
+                    }
+                });
+            } else {
+                // columnName with any alias and unique to one table
+                Table table = tables.iterator().next();
+                idExpr.setName(String.join(DOT, table.getAlias(), columnName));
+            }
+        });
+    }
+
+    private void visitTable(SQLQueryExpr root,
+                            Consumer<SQLExprTableSource> visit) {
+        root.accept(new MySqlASTVisitorAdapter() {
+            @Override
+            public void endVisit(SQLExprTableSource tableExpr) {
```

Comment:
What's the benefit of defining the function in this way?

Reply:
Ideally we would have needed a single Visitor like `TermFieldRewriter` and `NestedFieldRewriter` .  But since we are doing different things (conflicting) while visiting `SQLExprTableSource` in `visitTable` and `visitColumnName`, we need to have two Visitors and also they should be able to share the state `tableByFieldName` and `tableNameToAlias`. This pattern cleanly solves this problem and also makes it more readable by using intention revealing function names.